src/xrpld/app/tx/detail/OnChainCredentials.cpp

```diff
@@ -0,0 +1,454 @@
+    Copyright (c) 2024 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <xrpld/app/tx/detail/OnChainCredentials.h>
+
+#include <xrpld/core/TimeKeeper.h>
+#include <xrpld/ledger/ApplyView.h>
+#include <xrpld/ledger/View.h>
+#include <xrpl/basics/Log.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/Indexes.h>
+#include <xrpl/protocol/TxFlags.h>
+#include <xrpl/protocol/st.h>
+
+#include <chrono>
+
+namespace ripple {
+
+/*
+    OnChainCredentials
+    ======
+
+   A verifiable credentials (VC
+   https://en.wikipedia.org/wiki/Verifiable_credentials), as defined by the W3C
+   specification (https://www.w3.org/TR/vc-data-model-2.0/), is a
+   secure and tamper-evident way to represent information about a subject, such
+   as an individual, organization, or even an IoT device. These credentials are
+   issued by a trusted entity and can be verified by third parties without
+   directly involving the issuer at all.
+*/
+
+//------------------------------------------------------------------------------
+
+bool
+credentialCheckExpired(
+    std::shared_ptr<SLE const> const& sle,
+    NetClock::time_point const& closed)
+{
+    std::uint32_t const exp =
+        sle->isFieldPresent(sfExpiration) ? sle->getFieldU32(sfExpiration) : 0;
+    std::uint32_t const now = closed.time_since_epoch().count();
+    return now > exp;
+}
+
+// special check for deletion
+static bool
+credentialCheckNotExpired(
+    std::shared_ptr<SLE const> const& sle,
+    NetClock::time_point const& closed)
+{
+    if (!sle->isFieldPresent(sfExpiration))
+        return false;
+
+    std::uint32_t const exp = sle->getFieldU32(sfExpiration);
+    std::uint32_t const now = closed.time_since_epoch().count();
+
+    return now <= exp;
+}
+
+// ------- CREATE --------------------------
+
+NotTEC
+CredentialCreate::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureOnChainCredentials))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto& tx = ctx.tx;
+    auto& j = ctx.j;
+
+    auto const optSubj = tx[~sfSubject];
```

Comment:
This is a comment across the whole PR - can whole names be used instead of just part of the name, especially for `subject` and `issuer`? It makes it a bit easier to read, and the full names aren't that much longer.

Reply:
fixed